infra/cifuzz/filestore/gitlab/__init__.py

```diff
@@ -0,0 +1,121 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Empty filestore implementation for platforms that haven't implemented it."""
+import logging
+
+import shutil
+import os
+import json
+import tempfile
+
+import environment
+import http_utils
+import filestore
+
+# pylint:disable=no-self-use,unused-argument
+
+
+class GitlabFilestore(filestore.BaseFilestore):
+  """Implementation of BaseFilestore using Gitlab job artifacts. Relies on
+  having a PRIVATE-TOKEN supplied to access the Gitlab API as seems
+  the only way to download an artifact from another job"""
+
+  def __init__(self, config):
+    super().__init__(config)
+    self.downloaded = set()
+    self.artifacts_dir = environment.get('CFL_ARTIFACTS_DIR', "artifacts")
+    self.download_dir = environment.get('CFL_DOWNLOAD_DIR', "download")
+
+  def _copy_fromdir(self, src, name, reason):
+    destdir = os.path.join(self.config.workspace, self.artifacts_dir, reason,
+                           name)
+    logging.info('Uploading %s to artifacts to %s', reason, destdir)
+    shutil.copytree(src, destdir)
+
+  def upload_crashes(self, name, directory):
+    """Gitlab artifacts implementation of upload_crashes."""
+    self._copy_fromdir(directory, name, "crashes")
+
+  def upload_corpus(self, name, directory, replace=False):
+    """Gitlab artifacts implementation of upload_corpus."""
+    self._copy_fromdir(directory, name, "corpus")
+
+  def upload_build(self, name, directory):
+    """Gitlab artifacts implementation of upload_build."""
+    self._copy_fromdir(directory, name, "build")
+
+  def upload_coverage(self, name, directory):
+    """Gitlab artifacts implementation of upload_coverage."""
+    # would be neat if we could save CI_JOB_ID somewhere,
+    # without needing a token with write-access
+    self._copy_fromdir(directory, name, "coverage")
+
+  def _get_jobid(self, projpath, headers, reason):
+    """Get a specific job id for the latest succesful pipeline
+    with the specific job name."""
+    jobsurl = os.getenv('CI_API_V4_URL') + projpath + "/jobs?scope=success"
+    with tempfile.NamedTemporaryFile() as tmp_file:
+      if not http_utils.download_url(jobsurl, tmp_file.name, headers=headers):
+        logging.error('Failed downloading %s', jobsurl)
+        return None
+      # jq '.[] | select(.name=="clusterfuzzlite-corpus") | .id'
+      with open(tmp_file.name, encoding='ascii') as json_file:
+        data = json.load(json_file)
+        for j in data:
```

Comment:
Is there a better variable name we can use than `j`

Reply:
job ;-)